src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java

```diff
@@ -1,16 +1,63 @@
 package pl.wrapper.parking.facade.domain;
 
+import lombok.extern.slf4j.Slf4j;
 import org.springframework.stereotype.Service;
 import pl.wrapper.parking.facade.ParkingService;
+import pl.wrapper.parking.facade.client.NominatimClient;
+import pl.wrapper.parking.facade.dto.NominatimLocation;
+import pl.wrapper.parking.facade.exception.AddressNotFoundException;
+import pl.wrapper.parking.infrastructure.error.Result;
 import pl.wrapper.parking.pwrResponseHandler.PwrApiServerCaller;
+import pl.wrapper.parking.pwrResponseHandler.dto.ParkingResponse;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
 
 @Service
-record ParkingServiceImpl(PwrApiServerCaller pwrApiServerCaller) implements ParkingService {
-    // 1. endpointy:
-    // - endpoint dla każdego parametru, zwraca jeden, kilka, lub nic
-    // - research, closest address
-    // - most free spots, everything that has => free spots, eveything has any free spots
-    // -* error handling, try to use as few exceptions as possible (think of a good way to do it)
-    //   dokumentacja + swagger do enpointa/endpointów i testy jednostkowe + test integracyjny dla waszego
-    // enpointa/enpoiontów
+@Slf4j
+record ParkingServiceImpl(PwrApiServerCaller pwrApiServerCaller, NominatimClient nominatimClient) implements ParkingService {
+
+    @Override
+    public Result<List<ParkingResponse>> getAllParkings() {
+        return Result.success(pwrApiServerCaller.fetchData());
+    }
+
+    @Override
+    public Result<ParkingResponse> getClosestParking(String address) {
+        Optional<NominatimLocation> geoLocation = nominatimClient.search(address, "json").next().blockOptional();
+        return Result.success(
+                geoLocation.map(location -> {
+                    log.info("Geocoded address for coordinates: {} {}", location.latitude(), location.longitude());
+                    return findClosestParking(location, getAllParkings().getData());
+                })
+                .orElseThrow(() -> {
```

Comment:
Let's follow the result pattern, we shouldn't throw here

Reply:
Could you explain why? What is the benefit of using Result and manually convert it to json (there was a problem with converting dates/time to json using manual approach) rather than just return a response that would be converted **automatically** and **correctly** and handle the errors in GlobalExceptionHandler by wrapping them using ErrorWrapper.
I think that the only class that is useful in infrastructure/error folder is ErrorWrapper. Because if we will have a new error in our logic we need to add it into ParkingError interface first and then add a new case within switch in getInfoByError method. Also why should we return Result.failure which will be converted into ErrorWrapper -> Response<String> if in GlobalExceptionHandler the raised exception will be automatically converted into Response<ErrorWrapper>

Reply:
1. I'm not exactly sure what date/time would you like to convert when returning an Error. Even if so, adding the JavaTime module and specifying the date format should solve the problems - after all, we aren't really doing it manually and solely by ourselves.
2. Adding any error to the interface and creating a new switch case for it are exactly two lines of code. Writing an exception handler method as you'd like is about ten times that. So, I don't really understand why you would portray it as a demanding endeavour.
3. Yes, the outcome is more or less same in both cases in terms what the client would receive. However, is not finding the desired object an unusual situation? While I do agree with throwing when the Nominatim client doesn't respond, here, not so much. As such, my personal preference, which you may or may not agree with, is that we do not throw here and use the Result class. And it's not just my baseless whim - I assure you, handling it in the above-mentioned manner will be faster time-wise. Our switch is using finite pattern-matching, so the complexity is smaller than the mapping for the ExceptionHandling bean, not even including the time it takes to throw and catch the exception itself (which is the biggest difference maker). Although, it's still mainly because I and Aleksy decided to use such a solution.
Hope this solved your doubts. If not - feel free to ask for more detailed explanation. 👍